### PR TITLE
LocalStorage logic has been changed

### PR DESCRIPTION
**Pull request by helen508light**

### CODE REVIEW

File: src/geocoder/geocodingCache.ts [hunk 1]
```diff
@@ -40,31 +38,34 @@ module powerbi.extensibility.geocoder {
...
             this.maxCacheSize = maxCacheSize;
             this.maxCacheSizeOverflow = maxCacheSizeOverflow;
             this.localStorageService = localStorageService;
         }
 
-    /**
-    * Retrieves the coordinate for the key from the cache, returning undefined on a cache miss.
-    */
+        private minimizeKey(key: string): string {
+            return key.split(";").pop().slice(0, -1);
```
Comment (uve) [inline src/geocoder/geocodingCache.ts]: replace it by RegExp

File: src/geocoder/geocodingCache.ts [hunk 2]
```diff
@@ -73,16 +74,34 @@ module powerbi.extensibility.geocoder {
...
             }
             // Check local storage cache
-                pair = this.localStorageService.getData(key);
-                if (pair) {
+            const minimizesKey: string = this.minimizeKey(key);
+            const localStoragePromise: IPromise<string> = this.localStorageService.get(GeocodingCache.TILE_LOCATIONS);
+            localStoragePromise.then((value) => {
+                const parsedValue = JSON.parse(value);
+                if (!parsedValue)
+                    return undefined;
```
Comment (uve) [inline src/geocoder/geocodingCache.ts]: add braces here

File: src/geocoder/geocodingCache.ts [hunk 3]
```diff
@@ -133,7 +152,21 @@ module powerbi.extensibility.geocoder {
         }
 
         private registerInStorage(key: string, coordinate: IGeocodeCoordinate): void {
-            this.localStorageService.setData(key, { coordinate: coordinate });
+            const valuesObj = {};
+            const minimizeKey: string = this.minimizeKey(key);
+            valuesObj[minimizeKey] = {
+                "long": coordinate.longitude,
```
Comment (uve) [inline src/geocoder/geocodingCache.ts]: rename "long" to "lon"

File: src/geocoder/geocodingCache.ts (hunk 1 above)
Comment (uve) [reply src/geocoder/geocodingCache.ts]: rename method to: getShortKey

File: src/geocoder/geocodingCache.ts [hunk 4]
```diff
@@ -73,16 +74,34 @@ module powerbi.extensibility.geocoder {
                 return pair.coordinate;
             }
             // Check local storage cache
-                pair = this.localStorageService.getData(key);
-                if (pair) {
+            const minimizesKey: string = this.minimizeKey(key);
```
Comment (uve) [inline src/geocoder/geocodingCache.ts]: use shortKey name for the variable

File: src/globemap.ts [hunk 5]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
-                return this.getBingMapsServerMetadata()
-                    .then((metadata: BingResourceMetadata) => {
-                        tileCache = [];
-                        let urlTemplate = metadata.imageUrl.replace("{culture}", this.currentLanguage);
-                        for (let level: number = GlobeMap.initialResolutionLevel; level <= GlobeMap.maxResolutionLevel; ++level) {
-                            let levelTiles = this.generateQuadsByLevel(level, urlTemplate, metadata.imageUrlSubdomains);
-                            this.mapTextures.push(this.createTexture(level, levelTiles));
-                            tileCache.push(levelTiles);
+        private minimizeBingMetadata(tileCacheArray): any[] {
+            if (!tileCacheArray || !tileCacheArray.length)
```
Comment (uve) [inline src/globemap.ts]: add  braces here

File: src/globemap.ts [hunk 6]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
+                return [];
+
+            tileCacheArray.forEach((obj) => {
+                for (let key in obj) {
+                    if (obj.hasOwnProperty(key)) {
+                        const slpitedArray: string[] = obj[key].split("?")[0].split("/");
+                        let file: string = slpitedArray.pop().split(".")[0];
```
Comment (uve) [inline src/globemap.ts]: use RegExp to fine "file" and "domain"

File: src/globemap.ts [hunk 7]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
+        }
+
+        private extendBingMetadata(tileCacheArray): any[] {
+            const additionsParams: string = "g=6766&mkt=en-US&shading=hill";
+            const extension: string = "jpeg";
+            const protocol: string = "https://";
+            const firstDomains: string = "tiles.virtualearth.net";
+            const lastDomain: string = "ecn";
+
+            if (!tileCacheArray || !tileCacheArray.length)
```
Comment (uve) [inline src/globemap.ts]: add braces here

File: src/globemap.ts [hunk 8]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
-                    });
+                    }
+                }
+            });
+
+            return tileCacheArray;
+        }
+
+        private extendBingMetadata(tileCacheArray): any[] {
+            const additionsParams: string = "g=6766&mkt=en-US&shading=hill";
```
Comment (uve) [inline src/globemap.ts]: what is 6766 here? are you sure that it's a constant?

File: src/globemap.ts [hunk 9]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
+                    let urlTemplate = metadata.imageUrl.replace("{culture}", this.currentLanguage);
+                    for (let level: number = GlobeMap.initialResolutionLevel; level <= GlobeMap.maxResolutionLevel; ++level) {
+                        let levelTiles = this.generateQuadsByLevel(level, urlTemplate, metadata.imageUrlSubdomains);
+                        this.mapTextures.push(this.createTexture(level, levelTiles));
+                        tileCacheValue.push(levelTiles);
+                    }
+
+                    const minimizedTileCacheData: string = JSON.stringify(this.minimizeBingMetadata(tileCacheValue));
+                    this.localStorageService.set(GlobeMap.TILE_STORAGE_KEY, minimizedTileCacheData);
+                    this.localStorageService.set(GlobeMap.TILE_LANGUAGE_CULTURE, this.currentLanguage);
```
Comment (uve) [inline src/globemap.ts]: Put Culture first here, because it's small value. TILE_STORAGE_KEY might exceed available space.

File: src/globemap.ts [hunk 10]
```diff
@@ -721,33 +719,104 @@ module powerbi.extensibility.visual {
...
             }
         }
 
+        private initTextures(): JQueryPromise<{}> {
+            this.mapTextures = [];
+            const tileCulturePromise: IPromise<string> = this.localStorageService.get(GlobeMap.TILE_LANGUAGE_CULTURE);
+            let tileCachePromise: IPromise<string> = this.localStorageService.get(GlobeMap.TILE_STORAGE_KEY);
+            let deferred = $.Deferred();
+
+            tileCulturePromise.then((tileCultureValue: string) => {
```
Comment (uve) [inline src/globemap.ts]: Why tileCacheValues depend on tileCultureValue?  Can we storage tileCacheValues names with CultureKey?

File: package.json [hunk 11]
```diff
@@ -7,10 +7,10 @@
     "email": "pbicvsupport@microsoft.com"
   },
   "scripts": {
-    "postinstall": "bower install && pbiviz update 1.12.0",
+    "postinstall": "bower install && pbiviz update 2.2.0",
```
Comment (ignatvilesov) [inline package.json]: Why did you remove package-lock.json?

File: src/geocoder/geocodingCache.ts (hunk 2 above)
Comment (helen508light) [reply src/geocoder/geocodingCache.ts]: has already been added
